libraries/chain_kv/include/b1/session/shared_bytes.hpp

```diff
@@ -159,17 +159,15 @@ inline shared_bytes shared_bytes::previous() const {
       auto& val = buffer[index];
       if (val) {
          --val;
-         ++index; // ensure we increment any indexes past this one
          break;
       }
+      val = std::numeric_limits<unsigned char>::max();
```

Comment:
I'm sure I brought this up.

Reply:
https://github.com/EOSIO/eos/pull/9635#discussion_r517608924

Reply:
test

Reply:
@brianjohnson5972 Can you clarify the semantics of `previous`?

Reply:
(there are 2 different issue, @swatanabe-b1 brought up that next()/previous() are not just inverse logic of each other, but 5,0 is definitely > 4,2555)

The key (shared_bytes) is lexographically ordered, so "a" < "aa", "b" > "aa" (and remember lexigraphical ordering is based off of unsigned char, not char), so looking at a byte array: [0x0] < [0x0,0x0] < [0x0, 0x0, 0x0], and [0x0,0x0] < [0x0, 0x1], and [0x1, 0x0] > [0x0, 0xff].  Everyone with this so far?

So the use of next() was used to bound the "key space" that existed under a prefix.  So if you had the prefix [0x0, 0x0, 0x0], you called next to get [0x0, 0x0, 0x1], which you could use to identify any key that was outside your "key space", since [0x0, 0x0, 0x0, 0xff, 0xff, ...] would return true if you compared it as <= [0x0, 0x0, 0x1].  Does this make sense so far?  This matches next() so far. An example that may help is that for db intrinsics, the prefix is the contract, scope, and table that IS that table, laid out like [ contract [ MSB, MSB-1, ...., LSB], scope [ MSB, MSB-1, ..., LSB], table [ MSB, MSB -1, ..., LSB] ] so when you increment that LSB of the prefix, you are selecting the same contract and scope, but the very next higher table.

when you end in 0xff is where things are more ambiguous.  So next() was implemented as you just drop the last character if it is 0xff before you increment it, so [0x0, 0x0, 0xff] would return [0x0, 0x1]. The problem I have with this approach is that  it makes for more ambiguity.  If previous and next both always kept the same length, other than for all zeros or all 0xff, respectively, than they can be called on each other. But in terms of the "key space" I am now seeing that next() is correct, since if you had [0x0, 0x0, 0xff], you would want [0x0, 0x1], since if you got back [0x0, 0x1, 0x0], then if rocksdb returned a key of [0x0, 0x1], which is definitely outside your "key space" it would still be less than [0x0, 0x1, 0x0], thus proving I am wrong.

So the question is, can we come up with a previous() implementation that is analogous to next()?  Having it just pop the last character when it is going to overflow is not correct, since doing so would WIDEN the "key space" (as opposed to in next() where it narrows it) since you don't want [0x0, 0x1,0x0] to return [0x0, 0x0] for previous(), since [0x0, 0x0, 0x0] ... [0x0, 0x0, 0xff, 0xff ... 0xff] are all less than it. Not sure how to do this here. For a real corollary between next() and previous(), I think you would have to have something ending in 0x0, change it to 0xff and then add 0xff forever, and then I think the description of next() and previous() bounding the  "key space" would work, but obviously we cannot do that.

So would it be better to change the names, like next => end_of_key_space()?? (or something like that) and previous() become ??

Reply:
I am working on this, what would be the expected result on the case of {5, 0} ? I guess it also makes sense to return {4} for this case, but then next({4}) = {5}, so the method would not be the exact opposite of next. My guess it is that probably this method is currently working for the use cases where it is currently being used, but let me know if we require it to correct and if that's the case also the expected result for the example above.

Reply:
ultimately, what the use case needs is something that will return any string less than the current one, since this is being used as a <= iterator check (for the reverse iterator) so it just needs to be a string that would not satisfy ==. The important part is that it needs to not also be less than another potential key in the database.  So we could drop previous() and just have a method called something like truncate() which returns a new shared_bytes one character shorter than the original, but the name truncate() would make me think that it is truncating this shared_bytes, so the name would need to make it clear it was returning a new truncated shared_bytes.

Or instead the underlying data() and size()-1 could just be used to do the same where previous() is called.

Reply:
maybe replace previous by "get_lesser_key" or "get_lesser" and return the shared_bytes one character shorter than the original?

Reply:
`previous` was removed and a new method called `truncate_key` to replace its use case was added.